py/server/deephaven/update_graph.py

```diff
@@ -0,0 +1,222 @@
+#
+# Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+"""This module provides access to the Update Graph (UG)'s locks that must be acquired to perform certain
+table operations. When working with refreshing tables, UG locks must be held in order to have a consistent view of
+the data between table operations.
+"""
+
+import contextlib
+from functools import wraps
+from typing import Callable, Union
+
+import jpy
+
+from deephaven import DHError
+
+_JUpdateGraph = jpy.get_type("io.deephaven.engine.updategraph.UpdateGraph")
+
+auto_locking = True
+"""Whether to automatically acquire the Update Graph (UG) shared lock for an unsafe operation on a refreshing 
+table when the current thread doesn't own either the UG shared or the UG exclusive lock. The newly obtained lock will 
+be released after the table operation finishes. Auto locking is turned on by default."""
+
+
+def has_exclusive_lock(ug: Union[_JUpdateGraph, "Table", "PartitionedTable", "PartitionTableProxy"]) -> bool:
```

Comment:
Two problems:
1) The API is exposing Java objects to the user.
2) The type hint is referring to a private variable.

Reply:
Many places in this file.